dafni/main_dafni.py

```diff
@@ -0,0 +1,210 @@
+                                                          for x in args.data_path]) + "_results.json"
+
+        Path(args.output_path).parent.mkdir(exist_ok=True, parents=True)
+
+    else:
+
+        args.output_path = Path(args.output_path)
+
+        args.output_path.parent.mkdir(exist_ok=True, parents=True)
+
+    return args
+
+
+def read_variables(variables_path: Path) -> dict:
+    """
+    Function to read the variables.json file specified by the user
+    :param variables_path: A Path object of the user-specified file path
+    :returns:
+            - dict - A valid dictionary consisting of the causal tests
+    """
+    if not variables_path.exists() or variables_path.is_dir():
+
+        raise ValidationError(f"Cannot find a valid settings file at {variables_path.absolute()}.")
+
+    else:
+
+        with variables_path.open('r') as file:
+
+            inputs = json.load(file)
+
+        return inputs
+
+
+def validate_variables(data_dict: dict) -> tuple:
+    """
+    Function to validate the variables defined in the causal tests
+    :param data_dict: A dictionary consisting of the pre-defined variables for the causal tests
+    :returns:
+            - tuple - Tuple consisting of the inputs, outputs and constraints to pass into the modelling scenario
+    """
+    if data_dict["variables"]:
+
+        variables = data_dict["variables"]
+
+        inputs = [Input(variable["name"], eval(variable["datatype"])) for variable in variables if
+                  variable["typestring"] == "Input"]
+
+        outputs = [Output(variable["name"], eval(variable["datatype"])) for variable in variables if
+                   variable["typestring"] == "Output"]
+
+        constraints = set()
+
+        for variable, _inputs in zip(variables, inputs):
+
+            if "constraint" in variable:
+                constraints.add(_inputs.z3 == variable["constraint"])
+
+    return inputs, outputs, constraints
+
+
+def main():
+    """
+    Main entrypoint of the script:
+    """
+    args = get_args()
+
+    # Step 0: Read in the runtime dataset(s)
+
+    try:
+
+        data_frame = pd.concat([pd.read_csv(d) for d in args.data_path])
+
+        # Step 1: Read in the JSON input/output variables and parse io arguments
+
+        variables_dict = read_variables(args.variables_path)
+
+        inputs, outputs, constraints = validate_variables(variables_dict)
+
+        # Step 2: Set up the modeling scenario and estimator
+
+        modelling_scenario = Scenario(variables=inputs + outputs, constraints=constraints)
+
+        modelling_scenario.setup_treatment_variables()
+
+        estimators = {"LinearRegressionEstimator": LinearRegressionEstimator}
+
+        # Step 3: Define the expected variables
+
+        expected_outcome_effects = {
+            "Positive": Positive(),
+            "Negative": Negative(),
+            "NoEffect": NoEffect(),
+            "SomeEffect": SomeEffect()}
```

Comment:
Is there any way we can get round this hardcoding? This makes it pretty much impossible for a user to implement customisations, and is also the only blocker to having a single default `main.py` file (I think), now we're putting variables in a file.